### PR TITLE
[Core] Configure encrypting for payment gateway configuration

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | no |
| BC breaks?      | no |
| Related tickets | #7582 |
| License         | MIT |

### CODE REVIEW

**Comment by pamil** [comment]
Based on https://github.com/Payum/PayumBundle/pull/414.

**Comment by pamil** [comment]
Btw. can you add "fixes" before issue number in the opening comment so it will be automatically closed when this one gets merged?

File: src/Sylius/Bundle/CoreBundle/Form/Extension/PaymentMethodTypeExtension.php [hunk 1]
```diff
@@ -43,6 +70,30 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                    $decryptedConfigurationValue = openssl_decrypt(
+                        base64_decode($configuration),
+                        $this->encryptingAlgorithm,
+                        $this->encryptingSecret,
+                        OPENSSL_RAW_DATA,
+                        $this->encryptingInitializationVector
+                    );
+
+                    if (false === $decryptedConfigurationValue) {
+                        throw new \RuntimeException(sprintf('Encrypting failed for value "%s"', $configuration));
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Form/Extension/PaymentMethodTypeExtension.php]: "Encrypting" -> "Decrypting"

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml [hunk 2]
```diff
@@ -86,6 +86,8 @@ payum:
         config_storage:
             Sylius\Bundle\PayumBundle\Model\GatewayConfig: { doctrine: orm }
 
+    storage_encrypting: { algorithm: 'AES-128-CBC', secret: %secret%, initialization_vector: 'initialization16' }
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml]: `%secret%` -> `"%secret%"`

File: src/Sylius/Bundle/CoreBundle/Resources/config/services/form.xml [hunk 3]
```diff
@@ -69,6 +69,9 @@
             <tag name="form.type_extension" extended-type="Sylius\Bundle\OrderBundle\Form\Type\CartItemType" />
         </service>
         <service id="sylius.form.extension.type.payment_method" class="Sylius\Bundle\CoreBundle\Form\Extension\PaymentMethodTypeExtension">
+            <argument>%payum.storage.encrypting_algorithm%</argument>
+            <argument>%payum.storage.encrypting_initialization_vector%</argument>
+            <argument>%secret%</argument>
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/form.xml]: Shouldn't it be something like `%payum.storage.encrypting_secret%`?

Comment (inssein) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/form.xml]: I believe that is the framework (symfony) secret he is utilizing.

Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/form.xml]: Yes, but he sets it earlier in `payum.storage_encrypting.secret`.

File: src/Sylius/Bundle/AdminBundle/Resources/views/PaymentMethod/update.html.twig [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        <input type="hidden" name="_method" value="PATCH" />
+        {% if configuration.vars.templates.form is defined %}
+            {% include configuration.vars.templates.form %}
+        {% else %}
+            {{ form_widget(form) }}
+        {% endif %}
+        {% include '@SyliusUi/Form/Buttons/_update.html.twig' with {'paths': {'cancel': path(configuration.getRouteName('index'), configuration.vars.route.parameters|default({}))}} %}
+        {{ form_row(form._token) }}
+    {{ form_end(form, {'render_rest': false}) }}
+{% endblock %}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/PaymentMethod/update.html.twig]: Why do we need this template?

File: src/Sylius/Bundle/AdminBundle/Resources/views/PaymentMethod/Update/_form.html.twig [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{% import '@SyliusUi/Macro/flags.html.twig' as flags %}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/views/PaymentMethod/Update/_form.html.twig]: Why do we need this template?

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/payment_method.yml [hunk 6]
```diff
@@ -33,11 +33,26 @@ sylius_admin_payment_method_create:
             vars:
                 subheader: sylius.ui.manage_payment_methods_available_to_your_customers
                 templates:
-                    form: SyliusAdminBundle:PaymentMethod:_form.html.twig
+                    form: SyliusAdminBundle:PaymentMethod:Create/_form.html.twig
                 route:
                     parameters:
                         factory: $factory
 
+sylius_admin_payment_method_update:
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/payment_method.yml]: Why do we need this? :D

**Comment by pamil** [comment]
Based on Payum/Payum#635, fixes Payum/Payum#634.

**Comment by makasim** [comment]
So to make use of this feature Payum has to be updated up to 1.4.x and the bundle up to 2.2.x.

Here's how it could be configured.

```yaml
payum:
    # ...
    dynamic_gateways:
        encryption:
            defuse_secret_key: '%defuse.secert_key%'
        config_storage:
            # ....
```

P.S. Once enabled it will crypt currently not crypted configs

**Comment by pamil** [comment]
That's awesome @makasim, I'll try it out in a few minutes :)

**Comment by makasim** [comment]
@pamil this is how you can get  a valid defuse secret key

```php
<?php
use Defuse\Crypto\Key;

var_dump(Key::createNewRandomKey()->saveToAsciiSafeString());
```

**Pull request closed without merging**